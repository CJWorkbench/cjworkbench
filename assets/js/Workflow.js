// This is the main script for the Workflow view

import React from 'react'
import WorkflowNavBar from './WorkflowNavBar'
import Lesson from './lessons/Lesson'
import PropTypes from 'prop-types'
import WorkflowEditor from './WorkflowEditor'
import { connect } from 'react-redux'
import { Trans } from '@lingui/macro'

export function MaybeNotYourWorkflow (props) {
  if (!props.isReadOnly && !props.isAnonymous) {
    return null // it's your workflow
  }

  let suggestion = null
  if (props.isLoggedIn) {
    suggestion = (
      <h3 className='suggestion'>
        <Trans id='js.Workflow.suggestion.duplicateToSaveChanges'>Duplicate it to save your changes</Trans>
      </h3>
    )
  } else {
    suggestion = (
      <h3 className='suggestion'>
        <Trans id='js.Workflow.suggestion.signInToSaveChanges' description='The tag is a link to the login page'>
          <a href={`/account/login/?next=/workflows/${props.workflowId}`} className='action-button '>Sign in</a> to save your changes
        </Trans>
      </h3>
    )
  }

  let inner, className
  if (props.isAnonymous) {
    className = 'is-anonymous'
    inner = (
      <>
        <h3><Trans id='js.Workflow.isAnonymous'>Demo workflow</Trans> -</h3>
        <p className='message' />
        {suggestion}
      </>
    )
  } else if (props.isReadOnly) {
    className = 'is-read-only'
    inner = (
      <>
        <h3><Trans id='js.Workflow.isShared'>You are viewing a shared workflow</Trans></h3>
        <p className='message' />
        {suggestion}
      </>
    )
  }

  return (
    <div className={`not-your-workflow ${className}`}>{inner}</div>
  )
}

// ---- WorkflowMain ----

export class Workflow extends React.PureComponent {
  static propTypes = {
    api: PropTypes.object.isRequired,
    isReadOnly: PropTypes.bool.isRequired,
    isAnonymous: PropTypes.bool.isRequired,
    workflow: PropTypes.object.isRequired,
    lesson: PropTypes.object, // or undefined
    loggedInUser: PropTypes.object // undefined if no one logged in (viewing public wf)
  }

  render () {
    const { lesson } = this.props

    let className = 'workflow-root'
    if (this.props.lesson) {
      className += ' in-lesson'
    }
    if (this.props.isReadOnly) {
      className += ' read-only'
    }
    if (this.props.isAnonymous) {
      className += ' anonymous'
    }

    return (
      <main className={className}>
        {lesson ? <Lesson {...lesson} /> : null}

        <div className='workflow-container'>
          <WorkflowNavBar
            workflow={this.props.workflow}
            lesson={lesson}
            api={this.props.api}
            isReadOnly={this.props.workflow.read_only}
            loggedInUser={this.props.loggedInUser}
          />

          <WorkflowEditor
            api={this.props.api}
          />

          <MaybeNotYourWorkflow
            workflowId={this.props.workflow.url_id}
            isLoggedIn={!!this.props.loggedInUser}
            isReadOnly={this.props.isReadOnly}
            isAnonymous={this.props.isAnonymous}
          />
        </div>
      </main>
    )
  }
}

// Handles addModule (and any other actions that change top level workflow state)
const mapStateToProps = (state) => {
  return {
    workflow: state.workflow,
    loggedInUser: state.loggedInUser,
    isAnonymous: state.workflow.is_anonymous,
    isReadOnly: state.workflow.read_only
  }
}

export default connect(
  mapStateToProps
)(Workflow)
