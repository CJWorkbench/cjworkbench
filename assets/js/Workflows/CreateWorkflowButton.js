import React, { useState } from 'react'
import { csrfToken } from '../utils'

function CreateWorkflowButton ({ children }) {
  const [isSubmitted, setSubmitted] = useState(false)

  return (
    <form className='create-workflow' method='post' action='/workflows' onSubmit={() => setSubmitted(true)}>
      <input type='hidden' name='csrfmiddlewaretoken' value={csrfToken} />
      <button type='submit' disabled={isSubmitted}>
        {children}
      </button>
    </form>
  )
}
export default CreateWorkflowButton
