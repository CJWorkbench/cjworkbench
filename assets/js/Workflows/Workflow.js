import React from 'react'
import PropTypes from 'prop-types'
import WorkflowContextMenu from './WorkflowContextMenu'
import WorkflowMetadata from './WorkflowMetadata'

function Workflow ({ workflow, deleteWorkflow, duplicateWorkflow, openShareModal }) {
  return (
    <div className='workflow'>
      <a href={`/workflows/${workflow.id}`}>
        <div className='workflow-title'>{workflow.name}</div>
        <WorkflowMetadata
          workflow={workflow}
          openShareModal={openShareModal}
        />
      </a>
      <WorkflowContextMenu
        workflowId={workflow.id}
        duplicateWorkflow={duplicateWorkflow}
        deleteWorkflow={deleteWorkflow}
      />
    </div>
  )
}
Workflow.propTypes = {
  workflow: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  deleteWorkflow: PropTypes.func, // func(id) => undefined, or null if not allowed to delete
  duplicateWorkflow: PropTypes.func.isRequired, // func(id) => undefined
  openShareModal: PropTypes.func // func(id) => undefined
}
export default React.memo(Workflow)
