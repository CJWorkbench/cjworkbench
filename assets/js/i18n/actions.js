const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
