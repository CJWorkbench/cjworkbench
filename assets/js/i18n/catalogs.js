import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
