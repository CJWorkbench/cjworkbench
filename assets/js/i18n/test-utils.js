import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
