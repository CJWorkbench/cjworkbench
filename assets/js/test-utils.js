/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
