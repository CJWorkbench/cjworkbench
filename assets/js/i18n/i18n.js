import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
