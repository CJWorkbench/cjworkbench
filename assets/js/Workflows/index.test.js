/* globals afterEach, beforeEach, describe, expect, it, jest */
import React from 'react'
import { tick, okResponseMock } from '../test-utils'
import { mountWithI18n } from '../i18n/test-utils'
import { act } from 'react-dom/test-utils'
import { Workflows } from './index'
import Workflow from './Workflow'
import CreateWorkflowButton from './CreateWorkflowButton'

describe('Workflow list page', () => {
  const testWorkflows = {
    owned: [
      {
        id: 1,
        name: 'Cleaning',
        owner_name: 'Fred Frederson',
        public: true,
        is_owner: true,
        last_update: '2010-10-18T00:30:00'
      },
      {
        id: 2,
        name: 'Charting',
        owner_name: 'Fred Frederson',
        public: false,
        is_owner: true,
        last_update: '2010-10-18T00:20:00'
      },
      {
        id: 3,
        name: 'Analysis',
        owner_name: 'Fred Frederson',
        public: false,
        is_owner: true,
        last_update: '2010-10-18T07:45:00'
      }
    ],
    shared: [
      {
        id: 7,
        name: 'Messy data cleanup',
        owner_name: 'John Johnson',
        public: false,
        is_owner: false,
        last_update: '2010-10-18T00:30:00'
      },
      {
        id: 8,
        name: 'Document search',
        owner_name: 'Sally Sallerson',
        public: true,
        is_owner: false,
        last_update: '2010-10-18T00:45:00'
      }
    ],
    templates: [
      {
        id: 10,
        name: 'Demo 1',
        owner_name: 'Workbench',
        public: false,
        is_owner: false,
        last_update: '2010-10-18T00:30:00'
      }
    ]
  }

  const api = {
    duplicateWorkflow: jest.fn(),
    deleteWorkflow: okResponseMock()
  }
  const wrapper = (props) => mountWithI18n(
    <Workflows
      api={api}
      user={{ id: 1 }}
      {...props}
    />
  )

  let globalConfirm
  beforeEach(() => {
    globalConfirm = global.confirm
    global.confirm = jest.fn()
  })
  afterEach(() => {
    global.confirm = globalConfirm
  })

  it('renders correctly', () => {
    const w = wrapper({ workflows: testWorkflows })
    expect(w).toMatchSnapshot()

    // owned tab should have 3 workflows
    w.find('.nav-link').find("Trans[defaults='My workflows']").simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(3)
    // Make sure there is a context menu for each workflow
    expect(w.find('.tab-pane.active WorkflowContextMenu')).toHaveLength(3)
    // Make sure there is a metadata line for each workflow in the list
    expect(w.find('.tab-pane.active WorkflowMetadata')).toHaveLength(3)

    // shared tab should have 2 workflows
    w.find('.nav-link').find("Trans[defaults='Shared with me']").simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(2)

    // template tab should have 1 workflow1
    w.find('.nav-link').find("Trans[defaults='Recipes']").simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(1)
  })

  it('delete a workflow', (done) => {
    const w = wrapper({ workflows: testWorkflows })
    global.confirm.mockReturnValue(true) // pretend the user clicked OK
    // We've clicked delete and now we have to wait for everything to update.
    // see https://facebook.github.io/jest/docs/asynchronous.html
    setImmediate(() => {
      // Shared tab should start with 2 workflows and have 1 after delete
      w.update()
      expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(3)
      w.instance().deleteWorkflow(3)

      setImmediate(() => {
        w.update()
        expect(api.deleteWorkflow.mock.calls.length).toBe(1)
        expect(api.deleteWorkflow.mock.calls[0][0]).toBe(3)
        expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(2) // one fewer workflow
        done()
      })
    })
  })

  it('duplicates a workflow', async () => {
    // let 2 workflows load in user's shared tab, duplicate 1 and activeTab should get set
    // to owned list with +1 workflow
    const w = wrapper({
      workflows: testWorkflows,
      api: {
        ...api,
        duplicateWorkflow: jest.fn(() => Promise.resolve({
          id: 666,
          name: 'Copy of Visualization',
          owner_name: 'Paul Plagarizer',
          public: false
        }))
      }
    })

    // Owned list should start with 3 WFs, shared with 2
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(3)
    w.find('.nav-link').find("Trans[defaults='Shared with me']").simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(2)

    act(() => {
      w.find('.tab-pane.active').find(Workflow).first().prop('duplicateWorkflow')(7)
    })
    expect(w.prop('api').duplicateWorkflow).toHaveBeenCalledWith(7)
    await tick()
    w.update()

    // Expect owned tab to now have 4 workflows
    // should be a new item at the top of the owned list
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(4)
  })

  it('owned pane should have create workflow link when no workflows', () => {
    const workflows = { ...testWorkflows, owned: [] }
    const w = wrapper({ workflows })

    // Owned list should have no workflows, instead a create workflow link
    w.find('.nav-link').find("Trans[defaults='My workflows']").simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(0)
    expect(w.find('.tab-pane.active').find(CreateWorkflowButton)).toHaveLength(1)
  })

  it('shared and template panes should have a placeholder when no workflows', (done) => {
    const w = wrapper({
      workflows: {
        owned: testWorkflows.owned,
        shared: [],
        templates: []
      }
    })
    setImmediate(() => {
      w.update()
      w.find('.nav-link').find("Trans[defaults='Shared with me']").simulate('click')
      expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(0)
      expect(w.find('.tab-pane.active .placeholder')).toHaveLength(1)
      w.find('.nav-link').find("Trans[defaults='Recipes']").simulate('click')
      expect(w.find('.tab-pane.active').find(Workflow)).toHaveLength(0)
      expect(w.find('.tab-pane.active .placeholder')).toHaveLength(1)
      done()
    })
  })

  it('should sort properly by date and name', () => {
    const w = wrapper({ workflows: testWorkflows })

    // sort by date ascending
    w.find('.sort-menu DropdownToggle button').simulate('click')
    w.find('button[data-comparator="last_update|ascending"]').simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow).get(0).key).toEqual('2')
    expect(w.find('.tab-pane.active').find(Workflow).get(1).key).toEqual('1')
    expect(w.find('.tab-pane.active').find(Workflow).get(2).key).toEqual('3')

    // sort by date descending
    w.find('.sort-menu DropdownToggle button').simulate('click')
    w.find('button[data-comparator="last_update|descending"]').simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow).get(0).key).toEqual('3')
    expect(w.find('.tab-pane.active').find(Workflow).get(1).key).toEqual('1')
    expect(w.find('.tab-pane.active').find(Workflow).get(2).key).toEqual('2')

    // sort by name ascending
    w.find('.sort-menu DropdownToggle button').simulate('click')
    w.find('button[data-comparator="name|ascending"]').simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow).get(0).key).toEqual('3')
    expect(w.find('.tab-pane.active').find(Workflow).get(1).key).toEqual('2')
    expect(w.find('.tab-pane.active').find(Workflow).get(2).key).toEqual('1')

    // sort by name descending
    w.find('.sort-menu DropdownToggle button').simulate('click')
    w.find('button[data-comparator="name|descending"]').simulate('click')
    expect(w.find('.tab-pane.active').find(Workflow).get(0).key).toEqual('1')
    expect(w.find('.tab-pane.active').find(Workflow).get(1).key).toEqual('2')
    expect(w.find('.tab-pane.active').find(Workflow).get(2).key).toEqual('3')
  })

  it('should allow delete of owned workflows', () => {
    const w = wrapper({ workflows: testWorkflows })
    w.find('.tab-pane.active .context-button').at(0).simulate('click')
    expect(w.find('.tab-pane.active button.delete-workflow')).toHaveLength(1)
  })

  it('should not allow delete of shared-with-me workflows', () => {
    const w = wrapper({ workflows: testWorkflows })
    w.find('.nav-link').find("Trans[defaults='Shared with me']").simulate('click')
    w.find('.tab-pane.active .context-button').at(0).simulate('click')
    expect(w.find('.tab-pane.active button.delete-workflow')).toHaveLength(0)
  })

  it('should not allow delete of templates', () => {
    const w = wrapper({ workflows: testWorkflows })
    w.find('.nav-link').find("Trans[defaults='Recipes']").simulate('click')
    w.find('.tab-pane.active .context-button').at(0).simulate('click')
    expect(w.find('.tab-pane.active button.delete-workflow')).toHaveLength(0)
  })
})
