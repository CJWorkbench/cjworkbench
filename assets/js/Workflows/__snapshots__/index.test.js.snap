// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Examples correctly 1`] = `
<div>
  <div
    class="page"
  >
    <nav
      class="main-nav"
    >
      <header>
        <a
          href="/workflows"
        >
          <img
            alt="Workbench"
            src="undefinedimages/workbench-logo-with-white-text.svg"
          />
        </a>
        <button
          class="toggle-open"
          title="menu"
        >
          /app/assets/icons/menu.svg
        </button>
      </header>
      <div
        class="details"
      >
        <ul>
          <li
            class=""
          >
            <a
              href="/workflows"
            >
              My workflows
            </a>
          </li>
          <li
            class=""
          >
            <a
              href="/workflows/shared-with-me"
            >
              Shared with me
            </a>
          </li>
          <li
            class="open"
          >
            <a>
              Community workflows
            </a>
          </li>
          <li
            class="parent"
          >
            <a
              href="/lessons"
            >
              Tutorials
            </a>
          </li>
        </ul>
        <div
          class="user-paths"
        >
          <h3
            class="display-name"
          >
            Example User
          </h3>
          <ul>
            <li
              class="parent"
            >
              <a
                href="/settings/plan"
              >
                Plan & Billing
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Language
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Log Out
              </a>
              <form
                action="/account/logout/"
                class="hidden"
                method="post"
              >
                <input
                  name="csrfmiddlewaretoken"
                  type="hidden"
                  value=""
                />
                <input
                  name="next"
                  type="hidden"
                  value="/workflows/examples"
                />
                <input
                  type="submit"
                />
              </form>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <main
      class="workflows"
    >
      <header>
        <h1>
          Community workflows
        </h1>
      </header>
      <div
        class="workflow-list example"
      >
        <table>
          <thead>
            <tr>
              <th
                class="title"
              >
                <a
                  href="#"
                >
                  Title
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="owner"
              >
                Owner
              </th>
              <th
                class="privacy"
              >
                <a
                  href="#"
                >
                  Privacy
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="fetches-per-day"
              >
                <a
                  href="#"
                >
                  Auto updates
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="updated"
              >
                <a
                  href="#"
                >
                  Updated
                  /app/assets/icons/sort-descending.svg
                </a>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/10"
                >
                  Demo 1
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/10"
                >
                  Workbench
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/10"
                >
                  private
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/10"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/10"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
</div>
`;

exports[`render My workflows correctly 1`] = `
<div>
  <div
    class="page"
  >
    <nav
      class="main-nav"
    >
      <header>
        <a>
          <img
            alt="Workbench"
            src="undefinedimages/workbench-logo-with-white-text.svg"
          />
        </a>
        <button
          class="toggle-open"
          title="menu"
        >
          /app/assets/icons/menu.svg
        </button>
      </header>
      <div
        class="details"
      >
        <ul>
          <li
            class="open"
          >
            <a>
              My workflows
            </a>
          </li>
          <li
            class=""
          >
            <a
              href="/workflows/shared-with-me"
            >
              Shared with me
            </a>
          </li>
          <li
            class=""
          >
            <a
              href="/workflows/examples"
            >
              Community workflows
            </a>
          </li>
          <li
            class="parent"
          >
            <a
              href="/lessons"
            >
              Tutorials
            </a>
          </li>
        </ul>
        <div
          class="user-paths"
        >
          <h3
            class="display-name"
          >
            Example User
          </h3>
          <ul>
            <li
              class="parent"
            >
              <a
                href="/settings/plan"
              >
                Plan & Billing
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Language
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Log Out
              </a>
              <form
                action="/account/logout/"
                class="hidden"
                method="post"
              >
                <input
                  name="csrfmiddlewaretoken"
                  type="hidden"
                  value=""
                />
                <input
                  name="next"
                  type="hidden"
                  value="/workflows"
                />
                <input
                  type="submit"
                />
              </form>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <main
      class="workflows"
    >
      <header>
        <h1>
          My workflows
        </h1>
      </header>
      <form
        action="/workflows"
        class="create-workflow"
        method="post"
      >
        <input
          name="csrfmiddlewaretoken"
          type="hidden"
          value=""
        />
        <button
          type="submit"
        >
          Create Workflow
        </button>
      </form>
      <div
        class="workflow-list owned"
      >
        <table>
          <thead>
            <tr>
              <th
                class="title"
              >
                <a
                  href="#"
                >
                  Title
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="owner"
              >
                Owner
              </th>
              <th
                class="privacy"
              >
                <a
                  href="#"
                >
                  Privacy
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="fetches-per-day"
              >
                <a
                  href="#"
                >
                  Auto updates
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="updated"
              >
                <a
                  href="#"
                >
                  Updated
                  /app/assets/icons/sort-descending.svg
                </a>
              </th>
              <th
                class="actions"
              />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/3"
                >
                  Analysis
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/3"
                >
                  Fred Frederson
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/3"
                >
                  secret link
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/3"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/3"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
              <td
                class="actions"
              >
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    class="icon-button"
                    title="menu"
                  >
                    /app/assets/icons/context-menu.svg
                  </button>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/1"
                >
                  Cleaning
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/1"
                >
                  Fred Frederson
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/1"
                >
                  public
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/1"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/1"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
              <td
                class="actions"
              >
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    class="icon-button"
                    title="menu"
                  >
                    /app/assets/icons/context-menu.svg
                  </button>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/2"
                >
                  Charting
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/2"
                >
                  Fred Frederson
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/2"
                >
                  private
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/2"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/2"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
              <td
                class="actions"
              >
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    class="icon-button"
                    title="menu"
                  >
                    /app/assets/icons/context-menu.svg
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
</div>
`;

exports[`render Shared correctly 1`] = `
<div>
  <div
    class="page"
  >
    <nav
      class="main-nav"
    >
      <header>
        <a
          href="/workflows"
        >
          <img
            alt="Workbench"
            src="undefinedimages/workbench-logo-with-white-text.svg"
          />
        </a>
        <button
          class="toggle-open"
          title="menu"
        >
          /app/assets/icons/menu.svg
        </button>
      </header>
      <div
        class="details"
      >
        <ul>
          <li
            class=""
          >
            <a
              href="/workflows"
            >
              My workflows
            </a>
          </li>
          <li
            class="open"
          >
            <a>
              Shared with me
            </a>
          </li>
          <li
            class=""
          >
            <a
              href="/workflows/examples"
            >
              Community workflows
            </a>
          </li>
          <li
            class="parent"
          >
            <a
              href="/lessons"
            >
              Tutorials
            </a>
          </li>
        </ul>
        <div
          class="user-paths"
        >
          <h3
            class="display-name"
          >
            Example User
          </h3>
          <ul>
            <li
              class="parent"
            >
              <a
                href="/settings/plan"
              >
                Plan & Billing
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Language
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                Log Out
              </a>
              <form
                action="/account/logout/"
                class="hidden"
                method="post"
              >
                <input
                  name="csrfmiddlewaretoken"
                  type="hidden"
                  value=""
                />
                <input
                  name="next"
                  type="hidden"
                  value="/workflows/shared-with-me"
                />
                <input
                  type="submit"
                />
              </form>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <main
      class="workflows"
    >
      <header>
        <h1>
          Workflows shared with me
        </h1>
      </header>
      <div
        class="workflow-list shared-with-me"
      >
        <table>
          <thead>
            <tr>
              <th
                class="title"
              >
                <a
                  href="#"
                >
                  Title
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="owner"
              >
                Owner
              </th>
              <th
                class="privacy"
              >
                <a
                  href="#"
                >
                  Privacy
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="fetches-per-day"
              >
                <a
                  href="#"
                >
                  Auto updates
                  /app/assets/icons/nothing.svg
                </a>
              </th>
              <th
                class="updated"
              >
                <a
                  href="#"
                >
                  Updated
                  /app/assets/icons/sort-descending.svg
                </a>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/8"
                >
                  Document search
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/8"
                >
                  Sally Sallerson
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/8"
                >
                  public
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/8"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/8"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
            </tr>
            <tr>
              <td
                class="title"
              >
                <a
                  href="/workflows/7"
                >
                  Messy data cleanup
                </a>
              </td>
              <td
                class="owner"
              >
                <a
                  href="/workflows/7"
                >
                  John Johnson
                </a>
              </td>
              <td
                class="privacy"
              >
                <a
                  href="/workflows/7"
                >
                  private
                </a>
              </td>
              <td
                class="fetches-per-day"
              >
                <a
                  href="/workflows/7"
                />
              </td>
              <td
                class="updated"
              >
                <a
                  href="/workflows/7"
                >
                  <time>
                    10y ago
                  </time>
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
</div>
`;
